directord/components/builtin_workdir.py

```diff
@@ -99,40 +108,31 @@ def client(self, cache, job):
             self.log.critical(str(e))
             return None, traceback.format_exc(), False, None
         else:
-            update_info = "Directory {} OK".format(workdir)
-            outcome = True
-            stderr = None
-            info = ""
-            if user:
-                try:
-                    try:
-                        uid = int(user)
-                    except ValueError:
-                        uid = pwd.getpwnam(user).pw_uid
-                        info += " uid {} found from name {}.".format(uid, user)
-
-                    if group:
-                        try:
-                            gid = int(group)
-                        except ValueError:
-                            gid = grp.getgrnam(group).gr_gid
-                            info += " gid {} found from name {}.".format(
-                                gid, group
-                            )
-                    else:
-                        gid = -1
-                except KeyError as e:
-                    outcome = False
-                    stderr = (
-                        "Failed to set ownership properties."
-                        " USER:{} GROUP:{}".format(user, group)
+            update_info = f"Directory {workdir} OK"
+            task_outcome = True
+            task_stderr = ""
+            task_info = ""
+
+            if user or group:
+                if recursive:
+                    user_group = f"{user or ''}:{group or ''}".rstrip(":")
+                    stdout, stderr, outcome = self.run_command(
+                        f"chown -R {user_group} {workdir}"
                     )
-                    info += " {}".format(str(e))
+                    task_outcome = task_outcome and outcome
+                    task_stderr = stderr.decode()
+                    task_info = stdout.decode()
                 else:
-                    os.chown(workdir, uid, gid)
```

Comment:
I think the recursive call could use `os.walk` to run recursively, instead of descending into a shell call. While shutils is part of the standard lib, I feel like using it for things like this is a slippery slope, especially for something that we can easily accomplish in pure python.

Reply:
Well, running `os.walk` recursively will take a lot of time and resources, python will be much slower here. Especially when there are a lot of files in nested directories, it can reach seconds or more.
`shutil` here allows to skip all gid/uid calculation and prevent possible issues there, I think it's much easier to do, but don't insist here.

Reply:
that's what I did..
https://github.com/directord/directord/pull/395/files#diff-930632aa690ede348a8da0cb9eeafb25d56d8770910269ac474eb4cae52ebf73R130

Reply:
I mean use shutil.chown, without having to rely on actual bash for the recursive call.

Reply:
I dislike using shell in our components wherever possible. I feel if we're forcing the component to use shell we might as well just tell the user to `RUN` a command given they'd be equivalent.

Reply:
`shutil.chown` doesn't support recursion :shrug:
maybe to switch all to use shell?